scrapy/exporters.py

```diff
@@ -253,12 +259,23 @@ def _build_row(self, values):
     def _write_headers_and_set_fields_to_export(self, item):
         if self.include_headers_line:
             if not self.fields_to_export:
+                self._is_data_loss_warning_displayed = True
                 if isinstance(item, dict):
                     # for dicts try using fields of the first item
                     self.fields_to_export = list(item.keys())
                 else:
                     # use fields declared in Item
                     self.fields_to_export = list(item.fields.keys())
+            elif not self._fail_on_dataloss_warned:
+                if isinstance(item, dict):
```

Comment:
Can we reduce code duplication here by doing something like `fields = item.keys() if isinstance(item, dict) else item.fields.keys()` and do the following checks only once?

Reply:
Yeah, sorry I will remove the redundant if statements